src/Orleans/Messaging/Message.cs

```diff
@@ -64,7 +64,9 @@ public static class Metadata
         public const int LENGTH_HEADER_SIZE = 8;
         public const int LENGTH_META_HEADER = 4;
 
-        private readonly Dictionary<Header, object> headers;
+        public bool DisposeScheduled { get; set; }
```

Comment:
The dispose either happens in callback after this line - https://github.com/dotnet/orleans/pull/1946/files#diff-faf1fffc73efc52f0cc10c13764b9283R183, or in few other places. The conditions of detecting in which place the dispose will happen is quite complicated (it depends on message direction, the target placement, categories, etc. So I thought it will be much more simpler to introduce single variable responsible for this.